cron/cron.go

```diff
@@ -109,130 +106,242 @@ func New(opts Options) (api.Interface, error) {
-			ev, err := c.informer.Events()
-			if err != nil {
-				return err
+	c.wg.Add(2)
+	defer c.wg.Wait()
+
+	go func(ctx context.Context) {
+		defer c.wg.Done()
+		if err := c.leadership.Run(ctx); err != nil {
+			if errors.Is(err, context.Canceled) {
+				// Ignore context cancellation errors
+				return
+			}
+			select {
+			case errCh <- err:
+			case <-ctx.Done():
 			}
+			return
+		}
+	}(ctx)
+
+	engineCtx, engineCancel := context.WithCancel(ctx)
+	defer engineCancel()
+
+	go func(ctx context.Context) {
+		defer c.wg.Done()
+		for {
+			leadershipCtx, leadershipCancel := context.WithCancel(ctx)
+
+			select {
+			case <-ctx.Done():
+				leadershipCancel()
+				return
+			default:
+				c.lock.Lock()
+				c.restartingCh = make(chan struct{})
+				c.lock.Unlock()
+
+				engine, err := engine.New(engine.Options{
+					Log:                              c.log,
+					Key:                              c.key,
+					Partitioner:                      c.part,
+					Client:                           c.client,
+					TriggerFn:                        c.triggerFn,
+					CounterGarbageCollectionInterval: c.gcgInterval,
+				})
+				if err != nil {
+					leadershipCancel()
+					select {
+					case errCh <- err:
+					case <-ctx.Done():
+					}
+					return
+				}
 
-			for {
-				select {
-				case <-ctx.Done():
-					return ctx.Err()
-				case e := <-ev:
-					if err := c.queue.HandleInformerEvent(ctx, e); err != nil {
-						return err
+				ectx, err := c.leadership.WaitForLeadership(leadershipCtx)
+				if err != nil {
+					leadershipCancel()
+					select {
+					case errCh <- err:
+					case <-ctx.Done():
 					}
+					return
+				}
+
+				// Store the engine once ready
+				c.lock.Lock()
+				c.engine.Store(engine)
+				close(c.readyCh)
+				c.lock.Unlock()
+
+				// Run engine with leadership context
+				if err := engine.Run(ectx); err != nil {
+					leadershipCancel()
+					select {
+					case errCh <- err:
+					case <-ctx.Done():
+					}
+					return
+				}
+
+				// Restart engine loop
+				c.log.Info("Restarting engine due to leadership change")
+
+				c.lock.Lock()
+				close(c.restartingCh)
+				c.readyCh = make(chan struct{})
+				c.lock.Unlock()
+
+				// Check for cancellation again
+				if err := ctx.Err(); err != nil {
+					leadershipCancel()
+					select {
+					case errCh <- err:
+					case <-ctx.Done():
+					}
+					return
 				}
 			}
-		},
-		func(ctx context.Context) error {
-			defer close(c.closeCh)
-			if err := c.leadership.WaitForLeadership(ctx); err != nil {
-				return err
-			}
-			if err := c.informer.Ready(ctx); err != nil {
-				return err
-			}
+		}
+	}(engineCtx)
 
-			close(c.readyCh)
-			c.log.Info("cron is ready")
-			<-ctx.Done()
+	<-ctx.Done() // block until cron is done
+	c.log.Info("cron shutdown gracefully")
 
+	select {
+	case err := <-errCh:
+		if errors.Is(err, context.Canceled) {
+			// Ignore context cancellation errors
 			return nil
-		},
-	).Run(ctx)
+		}
+		return err
+	default:
+		return nil
+	}
+}
```

Comment:
The top level loop seems a bit confused.


```go
// Run is a blocking function that runs the cron instance.
func (c *cron) Run(ctx context.Context) error {
	if !c.running.CompareAndSwap(false, true) {
		return errors.New("cron already running")
	}

	defer close(c.closeCh)

	return concurrency.NewRunnerManager(
		c.leadership.Run,
		func(ctx context.Context) error {
			for {
				engine, err := engine.New(engine.Options{
					Log:         c.log,
					Key:         c.key,
					Partitioner: c.part,
					Client:      c.client,
					TriggerFn:   c.triggerFn,

					CounterGarbageCollectionInterval: c.gcgInterval,
				})
				if err != nil {
					return err
				}

				c.engine.Store(engine)

				ectx, err := c.leadership.WaitForLeadership(ctx)
				if err != nil {
					return err
				}

				c.lock.Lock()
				close(c.readyCh)
				c.lock.Unlock()

				err = engine.Run(ectx)

				c.lock.Lock()
				c.readyCh = make(chan struct{})
				c.lock.Unlock()

				if ctx.Err() != nil {
					c.log.Info("cron shutdown gracefully")
					return err
				}

				c.log.Info("Restarting engine due to leadership change")
			}
		},
	).Run(ctx)
}
```

Reply:
Im not sure what you mean by confused. I had to remove the runner manager bc when the leadership & engine runners would shut down to restart it would cause the cron itself to shutdown and exit as well bc its in runners and with the runner manager it kills all runners.

Reply:
The runner manager will only close the sub contexts which belong to that runner. The engine should be in a wrapped sub context to the leadership. Leadership change should not close the outer runner manager, only when there is an error or if cron is shutting down.